### PR TITLE
add "Preserve Interword Spacing" and "Page Seg Mode"

### PR DESCRIPTION
**Pull request by macb74**

### CODE REVIEW

**Review by dadoonet — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution.

It looks yet very good. I left some comments. Could you commit your new changes to your branch so the PR will be updated?

Also, I'd add a PR description, using the same content you wrote for the guide.

Would it be possible to add 2 unit tests for:

* `page_seg_mode`
* `preserve_interword_spacing`

For unit tests, you can add those at the end of [this `testOcr()` method](https://github.com/dadoonet/fscrawler/blob/master/tika/src/test/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaDocParserTest.java#L576).

File: docs/source/user/ocr.rst [hunk 1]
```diff
@@ -160,3 +166,28 @@ Supported strategies are:
...
+
+OCR Page Seg Mode(Tesseract)
+----------------
+
+Set Tesseract to only run a subset of layout analysis and assume a certain form of image. The options for N are:
+
+* ``0`` = Orientation and script detection (OSD) only.
+* ``1`` = Automatic page segmentation with OSD.
+* ``2`` = Automatic page segmentation, but no OSD, or OCR. (not implemented)<
+* ``3`` = Fully automatic page segmentation, but no OSD. (Default)
```
[the remark this answers is not in the document]
Comment (macb74) [reply docs/source/user/ocr.rst:178]: Here `1` is written: https://cwiki.apache.org/confluence/display/tika/TikaOCR

I am a bit confused, but I think documentation of lib is right.

File: settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Ocr.java [hunk 2]
```diff
@@ -34,6 +34,9 @@ public class Ocr {
     private boolean enabled = true;
     // Pdf OCR Strategy
     private String pdfStrategy = "ocr_and_text";
+    // PDF Page Seg Mode
+    private String pageSegMode = "1";
```
[the remark this answers is not in the document]
Comment (macb74) [reply settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Ocr.java:38]: Lib requires a String, so I iplemented a String. Sure, oly numbers are valid.

**Comment by macb74** [comment]
I have implemented the changes as suggested, or added appropriate comments.
I have never written unit tests so I am a bit lost here. Maybe you can support.

For `preserve_interword_spacing `we would need to store a new image that creates blanks at ocr. In the gegebprobe the blanks would be gone.

I still have no idea how to test `page_seg_mode` reasonably.

Regards Martin

File: settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Ocr.java [hunk 3]
```diff
@@ -34,6 +34,9 @@ public class Ocr {
...
+    private boolean preserveInterwordSpacing = false;
```
Comment (dadoonet) [inline settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Ocr.java:39]: I'd use a `Boolean` here with `null` as a default value.

File: tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java [hunk 4]
```diff
@@ -168,6 +168,10 @@ private static void initContext(Fs fs) {
                 TesseractOCRConfig config = new TesseractOCRConfig();
                 logger.debug("Tesseract Language set to [{}].", fs.getOcr().getLanguage());
                 config.setLanguage(fs.getOcr().getLanguage());
+                logger.debug("Tesseract PageSegMode set to [{}].", fs.getOcr().getPageSegMode());
+                config.setPageSegMode(fs.getOcr().getPageSegMode());
+                logger.debug("Tesseract preserveInterwordSpacing set to [{}].", fs.getOcr().getPreserveInterwordSpacing());
```
Comment (dadoonet) [inline tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:173]: Same here:

```java
if (fs.getOcr(). getPreserveInterwordSpacing() != null) {
  logger.debug("Tesseract preserveInterwordSpacing set to [{}].", fs.getOcr().getPreserveInterwordSpacing());
  config.setPreserveInterwordSpacing(fs.getOcr().getPreserveInterwordSpacing());
}
```

**Review by dadoonet — CHANGES_REQUESTED** [review]
Left some comments.

About the tests, I'll come later with some ideas.
But do you have some binary files we could use to test the behavior?

**Comment by macb74** [comment]
> Left some comments.
> 
> About the tests, I'll come later with some ideas. But do you have some binary files we could use to test the behavior?

Maybe we can use the binaries from [Using different Page Segmentation Modes](https://tesseract-ocr.github.io/tessdoc/Command-Line-Usage.html) -> Capture: Using different Page Segmentation Modes
Maybe we can also do some tests with these outputs.

File: docs/source/user/ocr.rst [hunk 5]
```diff
@@ -160,3 +166,29 @@ Supported strategies are:
 
 .. note:: When omitted, ``ocr_and_text`` value is used. If you have performance issues, it's worth using the ``auto`` option
 instead as only documents with barely no text will go through the OCR process.
+
+
+OCR Page Seg Mode
+----------------------------
```
Comment (dadoonet) [inline docs/source/user/ocr.rst:172]: Please make the length identical...

It should be:

```
OCR Page Seg Mode
-----------------
```

File: tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java [hunk 6]
```diff
@@ -168,6 +168,14 @@ private static void initContext(Fs fs) {
                 TesseractOCRConfig config = new TesseractOCRConfig();
                 logger.debug("Tesseract Language set to [{}].", fs.getOcr().getLanguage());
                 config.setLanguage(fs.getOcr().getLanguage());
+                if (fs.getOcr().getPageSegMode() != null) {
+                    logger.debug("Tesseract PageSegMode set to [{}].", fs.getOcr().getPageSegMode());
+                    config.setPageSegMode("" + fs.getOcr().getPageSegMode());
+                }
+                if (fs.getOcr().getPageSegMode() != null) {
```
Comment (dadoonet) [inline tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:175]: It should be:

```java
if (fs.getOcr().getPreserveInterwordSpacing() != null) {
```

File: docs/source/user/ocr.rst [hunk 7]
```diff
@@ -160,3 +166,29 @@ Supported strategies are:
...
+
+Set Tesseract to only run a subset of layout analysis and assume a certain form of image. The options for N are:
+
+* ``0`` = Orientation and script detection (OSD) only.
+* ``1`` = Automatic page segmentation with OSD. (Default)
```
Comment (dadoonet) [inline docs/source/user/ocr.rst:177]: Let's remove ` (Default)`

File: docs/source/user/ocr.rst (hunk 5 above)
Comment (dadoonet) [reply docs/source/user/ocr.rst:172]: It's not solved. Exact code should be:

```
OCR Page Seg Mode
-----------------
```

And not:

```
OCR Page Seg Mode
----------------------------
```

Comment (macb74) [reply docs/source/user/ocr.rst:172]: Sorry, forgot to commit.

**Comment by macb74** [comment]
I updated my code.

**Review by dadoonet — APPROVED** [review]
LGTM!

Sorry it took so long to merge it ;)
But many thanks for your contribution to the code base.

**Pull request merged**